tests/test_repo_interactions.py

```diff
@@ -127,8 +127,23 @@ def test_chartpress_run(git_repo, capfd):
     git_repo.git.stash("pop")
 
 
+    # verify usage of --publish-chart when the chart version exists in the chart
+    # repo already
+    out = _capture_output(
```

Comment:
For my curiosity: why do the tests use `_capture_output()` instead of [the `capfd` fixture](https://docs.pytest.org/en/latest/reference.html#std-fixture-capfd)?

Reply:
@betatim I made these tests without being very knowledgeable about pytest, perhaps the capfd could have been used instead but I wasn't aware. I'm still not =/

The implementation of the function is printed below. It contains quite a bit of custom logic.

```python
def _capture_output(args, capfd, expect_output=False):
    """
    Calls chartpress given provided arguments and captures the output during the
    call.
    """
    # clear cache of in memory cached functions
    # this allows us to better mimic the chartpress CLI behavior
    chartpress.image_needs_building.cache_clear()
    chartpress.image_needs_pushing.cache_clear()
    chartpress.latest_tag_or_mod_commit.cache_clear()

    # first flush past captured output, then run chartpress, and finally read
    # and save all output that came of it
    _, _ = capfd.readouterr()
    chartpress.main(args)
    out, err = capfd.readouterr()
    if not expect_output:
        assert out == ""

    # since the output was captured, print it back out again for debugging
    # purposes if a test fails for example
    header = f'--- chartpress {" ".join(args)} ---'
    footer = "-" * len(header)
    print()
    print(header)
    print("out:")
    print(out)
    print("err:")
    print(err, file=sys.stderr)
    print(footer)
    return err
```

Reply:
Aha! `_capture_output()` runs chartpress and captures the output using the normal pytest way. Because of the leading `_` I thought it was a pytest internal function we pulled in and used directly. Thanks for explaining!